bindings/python/tests/pire_test.py

```diff
@@ -0,0 +1,401 @@
+    pire.NonrelocScannerNoMask,
+    pire.SimpleScanner,
+    pire.SlowScanner,
+    pire.CapturingScanner,
+]
+
+
+def check_scanner(scanner, accepts=(), rejects=()):
+    for line in accepts:
+        assert scanner.Matches(line), '"%s"' % line
+    for line in rejects:
+        assert not scanner.Matches(line), '"%s"' % line
+
+
+def check_equivalence(scanner1, scanner2, examples):
+    for line in examples:
+        assert scanner1.Matches(line) == scanner2.Matches(line), '"%s"' % line
+
+
+def check_state(state, final=None, dead=None, accepted_regexps=None):
+    if dead is None and final:
+        dead = False
+
+    if accepted_regexps is None and final is False:
+        accepted_regexps = ()
+    if isinstance(state, (pire.SimpleScannerState, pire.CapturingScannerState)):
+        accepted_regexps = None
+
+    assert final is None or state.Final() == final
+    assert dead is None or state.Dead() == dead
+    assert accepted_regexps is None or tuple(state.AcceptedRegexps()) == tuple(accepted_regexps)
+
+
+@pytest.fixture(params=SCANNER_CLASSES)
+def scanner_class(request):
+    return request.param
+
+
+@pytest.fixture()
+def parse_scanner(scanner_class):
+    def scanner_factory(pattern, options=""):
+        lexer = pire.Lexer(pattern, options)
+        fsm = lexer.Parse()
+        return scanner_class(fsm)
+    return scanner_factory
+
+
+@pytest.fixture()
+def example_scanner(parse_scanner):
+    return parse_scanner("s(om)*e")
+
+
+def check_scanner_is_like_example_scanner(scanner):
+    check_scanner(scanner, accepts=["se", "somome"], rejects=["", "s"])
+
+
+class TestFsm(object):
+    def test_fsm_is_default_constructible(self):
+        f = pire.Fsm()
+        assert 1 == f.Size()
+
+    def test_fsm_can_be_made_false(self):
+        f = pire.Fsm.MakeFalse()
+        assert 1 == f.Size()
+
+    def test_default_fsm_compiles_to_default_scanner(self):
+        scanner = pire.Fsm().Compile()
+        assert pire.Scanner == type(scanner)
+
+    def test_fsm_compiles_to_scanner_of_choice(self, scanner_class):
+        assert scanner_class == type(pire.Fsm().Compile(scanner_class))
+
+    def test_fsm_is_copy_constructible(self):
+        fsm = pire.Fsm().Append("ab")
+        fsm_copy = pire.Fsm(fsm)
+        assert fsm_copy is not fsm
+        check_equivalence(
+            fsm.Compile(),
+            fsm_copy.Compile(),
+            ["", "a", "ab", "ab-", "-"]
+        )
+
+        fsm.Append("c")
+        assert not fsm_copy.Compile().Matches("abc")
+
+    def test_fsm_supports_appending_special(self, scanner_class):
+        fsm = pire.Fsm()
+        fsm.AppendSpecial(pire.BeginMark)
+        fsm.Append('a')
+        fsm.AppendSpecial(pire.EndMark)
+
+        scanner = scanner_class(fsm)
+        state = scanner.InitState()
+
+        check_state(state.Begin(), final=False, dead=False)
+        check_state(state.Run('a'), final=False, dead=False)
+        check_state(state.Step(pire.EndMark), final=True)
```

Comment:
`final=True` implies `dead=False`, and this logics is applied by `check_state` if `dead` argument is not specified or `None`.